substrate/frame/support/src/traits/members.rs

```diff
@@ -391,3 +394,141 @@ impl<T: Clone + Ord> ChangeMembers<T> for () {
 	fn set_members_sorted(_: &[T], _: &[T]) {}
 	fn set_prime(_: Option<T>) {}
 }
+
+/// Access data associated to a unique membership
+pub trait MembershipInspect<M, AccountId, MembershipId>
+where
+	M: Membership<Id = MembershipId>,
+	MembershipId: Parameter,
+{
+	type MembershipsIter: Iterator<Item = MembershipId>;
+
+	/// Retrieve membership data that is expected to belong to member
+	fn get_membership(id: impl Into<MembershipId>, member: &AccountId) -> Option<M>;
+
+	/// Retrieve all memberships belonging to member
+	fn account_memberships(member: &AccountId) -> Self::MembershipsIter;
+
+	/// Check membership is owned by the given account
+	fn has_membership(id: impl Into<MembershipId>, member: &AccountId) -> bool {
+		Self::get_membership(id, member).is_some()
+	}
+}
+
+/// Change data related to a unique membership
+pub trait MembershipMutate<M, AccountId, MembershipId>
+where
+	M: Membership<Id = MembershipId>,
+	MembershipId: Parameter,
+{
+	/// Update the membership possibly changing its owner
+	fn update(
+		id: impl Into<MembershipId>,
+		membership: M,
+		maybe_member: Option<AccountId>,
+	) -> DispatchResult;
+}
+
+/// A unique membership
+pub trait Membership: codec::Decode + codec::Encode {
+	type Id;
+
+	fn id(&self) -> &Self::Id;
+
+	fn has_expired(&self) -> bool {
+		false
+	}
+}
+
+/// A membership with a rating system
+pub trait RankedMembership<Rank = GenericRank>: Membership
+where
+	Rank: Eq + Ord,
+{
+	fn rank(&self) -> &Rank;
+	fn rank_mut(&mut self) -> &mut Rank;
+}
+
+/// A generic rank in the range 0 to 100
+#[derive(
+	Clone,
+	Copy,
+	Debug,
+	Default,
+	Eq,
+	Ord,
+	PartialEq,
+	PartialOrd,
+	codec::Decode,
+	codec::Encode,
+	codec::MaxEncodedLen,
+	scale_info::TypeInfo,
+)]
+pub struct GenericRank(u8);
+impl GenericRank {
+	pub const MIN: Self = GenericRank(0);
+	pub const MAX: Self = GenericRank(100);
+	pub const ADMIN: Self = Self::MAX;
+
+	pub fn set(&mut self, n: u8) {
+		*self = Self(n.min(Self::MAX.0))
+	}
+	pub fn promote_by(&mut self, n: NonZeroU8) {
+		*self = Self(self.0.saturating_add(n.get()).min(Self::MAX.0))
+	}
+	pub fn demote_by(&mut self, n: NonZeroU8) {
+		*self = Self(self.0.saturating_sub(n.get()).max(Self::MIN.0))
+	}
+}
+impl From<u8> for GenericRank {
+	fn from(value: u8) -> Self {
+		Self(value)
+	}
+}
+
+const MEMBERSHIP_INFO_ATTRIBUTE: [u8; 10] = *b"memberinfo";
```

Comment:
What do you think about adding `MEMBERSHIP_INFO_ATTRIBUTE` to  the [`PalletAttributes` enum](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/nfts/src/types.rs#L355)?
Maybe we should extract that enum to a common space and rename to `NftSystemAttributes`..

Reply:
Extracting it somewhere in frame-support would be better :) we shouldn't create a dependency with the nfts pallet which is also why I made the ItemConfig an associated type.